### ISSUE TITLE
imageio v3 api reads all pages of a multipage TIFF at once

### ISSUE THREAD

**Issue by thvoigtmann** [imageio/imageio#818]
I have a two-page TIFF, one 8bit grayscale image of shape (15,10) per page, and I am expecting `imageio.v3.imiter` to iterate over the two pages. Yet, it reads them both at once:
```python
>>> from skimage import data_dir
>>> file = data_dir + '/multipage.tif'
>>> import imageio
>>> len([*imageio.v3.imiter(file)])
1
>>> for img in imageio.v3.imiter(file):
...     img.shape
... 
(2, 15, 10)

```

The same code prior to [#805](https://github.com/imageio/imageio/pull/805) fulfills my expectation and iterates over the two pages:
```python
>>> from skimage import data_dir
>>> file = data_dir + '/multipage.tif'
>>> import imageio
>>> len([*imageio.v3.imiter(file)])
2
>>> for img in imageio.v3.imiter(file):
...     img.shape
... 
(15, 10)
(15, 10)
```

The TIFF file in question is part of `skimage`: https://github.com/scikit-image/scikit-image/blob/c95a8fd46a49a38de1f6032c84d9d61160b78283/skimage/data/multipage.tif
It appears to have been created with ImageMagick.

**Comment by thvoigtmann** [comment]
@FirefoxMetzger It doesn't appear to be specific to that TIFF file, nor to whether it's grayscale or RGB. I attach a little TIFF that I made with the GIMP by exporting two layers, each of shape (15, 10, 3). Same story as above, with the newest `imageio.v3.imiter` I get a single array of shape (2, 15, 10, 3).
File in a ZIP archive (supported by github): [multipage.zip](https://github.com/imageio/imageio/files/8830673/multipage.zip)

**Comment by FirefoxMetzger** [comment]
Sorry for taking such a long time to get around to the issue @thvoigtmann . This month doesn't leave me much time to work on side projects.

@almarklein perhaps you have some time to investigate this? Otherwise, it may have to wait until I am a little more free.

**Comment by almarklein** [comment]
Now that I think of it, I think this can be explained by the addition of the "?"in #805. This means that without specifying whether it's an image or volume, it assumes that one is interested in the full array:

https://github.com/imageio/imageio/blob/aba1eecb29cbbbfbd688ccba878c67d53b59204a/imageio/plugins/tifffile.py#L428-L449

So what I think happened is that in moving from `v2.volread` to `v3.imread` you lose the info that a volume is intended, which breaks certain use-cases, and was reported in #802. This was then fixed in #805 by assuming things to be a volume in more cases. But this caused the problem reported in this issue.

I'm not really sure what a good solution would be, but I think one way or another the user must be able to specify that an image or volume is assumed.

**Comment by almarklein** [comment]
@FirefoxMetzger good luck with your thesis!! :)

**Comment by FirefoxMetzger** [comment]
> So what I think happened is that in moving from v2.volread to v3.imread you lose the info that a volume is intended,

I would argue that we can get that information from the particular flavor of the TIFF. tifffile has flags for the different flavors it supports and, depending on that we should be able to decide what shape should be returned. Some of the relevant flags are [listed here (and following)](https://github.com/cgohlke/tifffile/blob/master/tifffile/tifffile.py#L8344) (I can't send a highlighted link, because my GH's UI breaks down for large files it seems).

I changed the behavior of `?` from `pages` to `series` in https://github.com/imageio/imageio/pull/805 because I assumed that series would auto-magically "do the right thing", but that is apparently not the case. This was to solve https://github.com/imageio/imageio/issues/802 where MetaMorph files (yet another TIFF variant) were read with the wrong shape by default. Here, a volume is expected, but it was read as individual pages. (Note that we don't have a regression test for this issue, because we lack a test image due to licensing constraints.)

I guess the proper solution for this is to write some branching code that sorts out which shape and type should be used depending on which flavor of TIFF the user tries to open. 

---

If my understanding is correct though, this would be a larger project so I can't say when I will get around to this. My thinking was that the license discussions (https://github.com/imageio/imageio/issues/809) appear like a more urgent issue to solve, so I will focus on that first.

**Comment by FirefoxMetzger** [comment]
I found myself with some spare time this morning, so I took a closer look at what is happening here.

The short version is that the behavior prior to ImageIO v2.19.3 is actually a bug. TIFF files that contain volumetric data, i.e., files which say in their metadata that the pages within should be interpreted as a volume, _should_ be read as a volume. However, prior to v2.19.3 this information was ignored and the file was simply read page by page.

That being said, I understand the desire/use-case to "ignore any metadata and just read pages individually". It would be cool if we can support this, though we would have to discuss how that would fit into the existing API (keeping in mind other formats).

> I'm not really sure what a good solution would be, but I think one way or another the user must be able to specify that an image or volume is assumed.

@almarklein In my opinion, a good solution would be to inspect the file's metadata (as we do now) and decide based on that. Most TIFF flavors store their shape information (volume, flat, how many images, ...) inside the file. tifffile can read that and, based on this information we should be able to decide the correct action auto-magically. This is consistent with our behavior on other formats (where we do the same), and for those it works really well. 

That said, for TIFF specifically there seems to be utility in circumventing this mechanism (i.e. "just read pages"). An elegant solution for this would be to add a custom plugin API for this, similar to what I am proposing in https://github.com/imageio/imageio/pull/853 . For this to work, however, we would have to first write a V3 plugin for tifffile ... which we should do anways, considering the amount of new features that tifffile now has and that we currenly don't support ...

---

Below are some details on what is going on and why the current behavior is desirable.

The Image that @thvoigtmann shared (`multipage3.tiff`) contains a multi-page image, i.e., a single image stored across multiple TIFF pages. Depending on context, this means that it is either a flat channel-first image or a volumetric image. Here it is probably a volume since each page is RGB. This is based on the file's metadata (it sets `NewSubfileType` on each page). Consequentially, ImageIO treats the file as containing a single ndimage and will yield that volume when reading the first (and only) image in the file via `iio.imread` and iterating over the image via `iio.imiter` will yield exactly one (nd)image. The same is true for the TIFF from scikit-image.

Here is the code that investigates the above:

```python
import tifffile

with tifffile.TiffFile("multipage3.tif") as file:
    assert file.is_multipage  # file contains a multi-page image/volume

    for page in file.pages:
        # Note: FILETYPE.PAGE means the page is part of a multi-page image
        assert page.subfiletype == tifffile.FILETYPE.PAGE

    # consequentially, the result is a single image where all pages
    # are stacked along the first dimension
    assert len(file.series) == 1
    assert file.series[0].shape == (2, 15, 10, 3)

assert iio.imread("multipage3.tif").shape == (2, 15, 10, 3)

imgs = [x for x in iio.imiter("multipage3.tif")]
assert len(imgs) == 1
```

To put this into perspective, we can use tifffile to create a TIFF that contains multiple independent images. The result will be a TIFF file in "shaped" flavor. Since we have stored two images in the file, iterating over the image via `iio.imiter` will yield two images:

```python
import tifffile
import numpy as np

img = np.full((35, 73, 3), 114, dtype=np.uint8)
with tifffile.TiffWriter("multipage_flat.tiff") as file:
    file.write(img)
    file.write(img)

with tifffile.TiffFile("multipage_flat.tiff") as file:
    assert file.is_shaped  # assert TIFF flavor

    # subfiletype is not used
    for page in file.pages:
        assert page.subfiletype == tifffile.FILETYPE.UNDEFINED

    # metadata is stored separately
    assert tuple(file.shaped_metadata[0]["shape"]) == (35, 73, 3)

    # tifffile deals with this internally and we can simply inspect high-level properties
    assert len(file.series) == 2
    assert file.series[0].shape == (35, 73, 3)
    assert file.series[1].shape == (35, 73, 3)

assert iio.imread("multipage_flat.tiff", index=0).shape == (35, 73, 3)
assert iio.imread("multipage_flat.tiff", index=1).shape == (35, 73, 3)

imgs = [x for x in iio.imiter("multipage_flat.tiff")]
assert len(imgs) == 2
```

This behavior is important, because TIFFs may combine volumetric images, flat images, or images in different shapes and sizes. This is messy, but if you need it, you need it, and we should handle this correctly:

```python
import tifffile
import numpy as np

volumetric = np.full((4, 255, 255, 3), 114, dtype=np.uint8)
flat = np.full((255, 255, 3), 114, dtype=np.uint8)
different_shape = np.full((120, 73, 3), 114, dtype=np.uint8)
with tifffile.TiffWriter("nightmare.tiff") as file:
    file.write(volumetric)
    file.write(flat)
    file.write(different_shape)

with tifffile.TiffFile("nightmare.tiff") as file:
    # subfiletype is not used
    for page in file.pages:
        assert page.subfiletype == tifffile.FILETYPE.UNDEFINED

    # the image is again "shaped"
    assert file.is_shaped

    # metadata is extracted correctly
    assert len(file.series) == 3
    assert file.series[0].shape == (4, 255, 255, 3)
    assert file.series[1].shape == (255, 255, 3)
    assert file.series[2].shape == (120, 73, 3)

# imread will read the image at the respective index
assert iio.imread("nightmare.tiff", index=0).shape == (4, 255, 255, 3)
assert iio.imread("nightmare.tiff", index=1).shape == (255, 255, 3)
assert iio.imread("nightmare.tiff", index=2).shape == (120, 73, 3)

# imiter will yield the three images in order
shapes = [(4, 255, 255, 3), (255, 255, 3), (120, 73, 3)]
for image, shape in zip(iio.imiter("nightmare.tiff"), shapes):
    assert image.shape == shape
```

@thvoigtmann @almarklein Does this shed some light on what is going on, or did this end up making the situation more confusing?

**Comment by almarklein** [comment]
> In my opinion, a good solution would be to inspect the file's metadata (as we do now) and decide based on that.  [...] That said, for TIFF specifically there seems to be utility in circumventing this mechanism (i.e. "just read pages"). An elegant solution for this would be to add a custom plugin API for this

Agreed!

**Comment by lschr** [comment]
Recent changes break reading individual frames from files written by `tifffile` with `contiguous=True`

```python
with tifffile.TiffWriter("img.tif") as tw:
    for f in frames:
        tw.write(f, contiguous=True)
```

which used to be the default until not too long ago. Thus, older files cannot be loaded frame by frame anymore. The same holds true for files created using various commercial microscopy software. The culprit is #805, which could be easily worked around using `mode="I"` until recently (#816).

Could #816 be reverted, as it removes an absolute essential feature without (IMHO) noticeable gain? 

Also, I would not expect all software to write correct metadata, so some way to force loading of individual frames may be a good thing to have in the long run anyways.

**Comment by FirefoxMetzger** [comment]
Making a long story short, there is good reasons to have `index=n` refer to the nth-series inside the TIFF file, but there are also good reasons to have `index=n` refer to the n-th page. The problem is that we can't have both in the old v2 API.

So, to resolve this (and some other issues related to this plugin), I've started work on a v3 plugin for tifffile (https://github.com/imageio/imageio/pull/914) 🚀 With it we can finally accommodate both use-cases: `index=n` can refer to the n-th page inside the file and `series=n` can refer to the n-th series inside the file.

@Ishr if you want to test this early, feel free to have a look at the PR :) I can't give an exact PR because I am juggling a few things in parallel right now, but I promise to incorporate feedback before it gets merged ;)

> Recent changes break reading individual frames from files written by tifffile with contiguous=True

I think this expected behavior, no? If you write frames with [`contiguous=True`](https://github.com/cgohlke/tifffile/blob/a73812e98f58b068a590de862ed547820fd332ad/tifffile/tifffile.py#L1783-L1793) you are saying that all pages are part of the same series. For TIFF's in `shaped` flavor this means that the result will be an ndimage of shape `(len(frames), *frame.shape)`. You can verify this by inspecting the resulting file's metadata:

```python
filename = tmp_path / "test.tiff"
data = np.full((128, 128, 3), 42, np.uint8)

with tifffile.TiffWriter(filename) as tw:
    tw.write(data, contiguous=True)
    tw.write(data, contiguous=True)
    tw.write(data, contiguous=True)
    tw.write(data, contiguous=True)

# note: using the v3 tiffile plugin for convenience :)
metadata = iio.immeta(filename)
assert metadata["is_shaped"] == True
assert metadata["shape"] == [4, 128, 128, 3]
```

So the expected result _should_ be that you get *one* ndimage instead of a sequence of frames. If you want a sequence of images, you could create your file as such (by dropping the `contiguous=True`).

> [...] which could be easily worked around using mode="I" until recently (https://github.com/imageio/imageio/pull/816).
 Could https://github.com/imageio/imageio/pull/816 be reverted, as it removes an absolute essential feature without (IMHO) noticeable gain?

The short answer is unfortunately no because we are moving away from the image modes `iIvV?`. 

The PR is a first step towards removing references to `image_mode` from our codebase so that I can eventually take it out completely. This then gives me some wiggle room to start refactoring and generalizing `Request` (https://github.com/imageio/imageio/issues/798) to allow support for cool things like video streams, IP cameras, reading/writing from/to GCP or AWS (similar to reading from HTTPS), etc. This is desirable, but difficult, because `Request` is very core to ImageIO and has become a bit non-modular over the years, so the first step is a series of reactors, one of which is #816 .

> Also, I would not expect all software to write correct metadata, so some way to force loading of individual frames may be a good thing to have in the long run anyways.

Yes, I agree it would be nice to have a way to read individual pages in a TIFF file. In https://github.com/imageio/imageio/pull/914 this can be done by using `iio.imread(..., index=int, plugin="tifffile")` or, if you want to iterate over all pages irrespective of the series they belong to you can do:

```python
with iio.imopen(filename, "r", plugin="tifffile") as file:
    for page in file.iter_pages():
        pass  # page is a ndarray with the data of that page
```

(Note that this is subject to change based on feedback I get on the PR :) )

**Comment by lschr** [comment]
@FirefoxMetzger Thanks for the reply!

> So the expected result _should_ be that you get _one_ ndimage instead of a sequence of frames.

I guess you are right. However, since `contiguous=True` used to be the default (and I wasn't even aware of this option), I have now a lot of old files with all images in a single series, and I guess I am not the only one. This was just one example illustrating why I find it useful to be able to read individual pages. The other, more concerning one is commercial software over which we have no control writing all images into a single series.

> The short answer is unfortunately no because we are moving away from the image modes `iIvV?`

I wasn't thinking about permanently reverting #816, but only until there is another way to load single pages (i.e., #914 is merged). I really appreciate your work moving imageio forward!

> if you want to test this early, feel free to have a look at the PR :) I can't give an exact PR because I am juggling a few things in parallel right now, but I promise to incorporate feedback before it gets merged ;)

Cool, I will have a look.
